### PR TITLE
chore(ci): skip regression checks for PRs with the `ci-condition: skip-regression` label

### PR DESCRIPTION
**Pull request by pront**
### Summary
* PR with the `"ci-condition: skip-regression"` label will skip the regression detector suite 
* PRs created by [dependabot[bot]](https://github.com/vectordotdev/vector/commits?author=dependabot%5Bbot%5D) will always have this label.
* What about regressions introduced by crate bumps? 
  * To mitigate this risk, we will now run the suite weekly. This will allows us to investigate any regressions before we release a new Vector versions.
  * Also, we can keep on eye for changelogs describing performance changes and trigger the regression suite on those PRs with review comments. 

In the future, we can also consider making this only run on schedule (e.g. nightly). But let's see how this smaller change works out.
  
## Testing

Check what happens when:
- A new commit is pushed
- A `/ci-run-regression` review comment is posted

### CODE REVIEW

File: .github/workflows/regression.yml [hunk 1]
```diff
@@ -32,7 +32,9 @@ on:
   # Don't want to run this on each PR commit, but because GH doesn't allow specifying different required checks
   # for pull request and merge queue, we need to "run" it in pull request, but in the jobs we will just auto pass.
   pull_request:
-
+  # Also, this workflow runs every Monday at 5:00 AM UTC.
+  schedule:
+    - cron: "0 5 * * 1"
```
Comment (jszwedko) [inline .github/workflows/regression.yml:37]: I'm a little concerned that a weeks worth of commits may make bisecting regressions more difficult, but we can see. We could change this to be nightly now rather than in the future.

File: .github/workflows/regression.yml [hunk 2]
```diff
@@ -754,8 +756,20 @@ jobs:
     env:
       FAILED: ${{ contains(needs.*.result, 'failure') }}
     steps:
+      # Check if the actor is Dependabot and skip if true, unless it's a review
```
Comment (jszwedko) [inline .github/workflows/regression.yml:759]: Rather than looking for `dependabot[bot]` could we instead look for a label like `ci/skip-regression`? I think that would have the advantage of letting us add the tag to PRs that we want to have skip the regression tests (e.g. docs changes).

You can change the labels that dependabot updates by default with: https://github.com/vectordotdev/vector/blob/441bf23e590fcc4f0043657beca25318f8c16cf5/.github/dependabot.yml#L8-L10

**Review by jszwedko — COMMENTED** [review]
A couple of minor comments, but otherwise 👍

File: .github/workflows/regression.yml (hunk 1 above)
Comment (pront) [reply .github/workflows/regression.yml:37]: Discussed offline, we will start with weekly runs. If this causes difficult investigations in the future, we will increase the frequency of this.

File: .github/dependabot.yml [hunk 3]
```diff
@@ -8,6 +8,7 @@ updates:
     labels:
       - "domain: deps"
       - "no-changelog"
+      - "ci-condition: skip-regression"
```
Comment (jszwedko) [inline .github/dependabot.yml:11]: I forgot there are multiple "ecosystems" defined in this file. I think this label needs to be added for each of them.

**Review by jszwedko — COMMENTED** [review]
One more comment, otherwise 👍

File: .github/dependabot.yml (hunk 3 above)
Comment (pront) [reply .github/dependabot.yml:11]: Thanks, added in https://github.com/vectordotdev/vector/pull/21507/commits/b9c7daee1676c76e65f8803ba2a1c143df188499

**Review by jszwedko — APPROVED** [review]

File: .github/workflows/regression.yml [hunk 4]
```diff
@@ -773,16 +777,17 @@ jobs:
...
-           echo "To trigger a run, leave a comment with \`/ci-run-regression\`" >> $GITHUB_STEP_SUMMARY
+          echo "### Workflow skipped" >> $GITHUB_STEP_SUMMARY
+          echo "This workflow doesn't run on PRs automatically." >> $GITHUB_STEP_SUMMARY
+          echo "This workflow doesn't run on PRs created by Dependabot." >> $GITHUB_STEP_SUMMARY
+          echo "To trigger a run, leave a review comment with \`/ci-run-regression\`" >> $GITHUB_STEP_SUMMARY
 
       - name: exit
         run: |
           echo "failed=${{ env.FAILED }}"
-          if [[ "$FAILED" == "true" ]] ; then
```
Comment (jszwedko) [inline .github/workflows/regression.yml:785]: Was this change necessary? The previous line could probably be deleted if so.

**Comment by pront** [comment]
Note: Squashed all 4 commits and forced push.

File: .github/workflows/regression.yml (hunk 4 above)
Comment (pront) [reply .github/workflows/regression.yml:785]: No, not necessary. In fact I will revert it since it's not relevant to this PR.

Comment (pront) [reply .github/workflows/regression.yml]: [revert unrelated change](https://github.com/vectordotdev/vector/pull/21507/commits/f4064364f4121bf95d6be779b2896c4dff39b0a2)

File: .github/workflows/regression.yml [hunk 5]
```diff
@@ -753,9 +755,11 @@ jobs:
       - analyze-experiment
     env:
       FAILED: ${{ contains(needs.*.result, 'failure') }}
+      SKIP_REGRESSION: "${{ contains(github.event.pull_request.labels.*.name, 'ci-condition: skip-regression') }}"
+
     steps:
       - name: (PR review) Submit PR result as failed
-        if: github.event_name == 'pull_request_review' && env.FAILED == 'true'
+        if: github.event_name == 'pull_request_review' && env.FAILED == 'true' && env.SKIP_REGRESSION == 'true'
```
Comment (jszwedko) [inline .github/workflows/regression.yml:762]: I'm a bit confused by this and the below change. We want to set the commit status if it failed _and_ the PR was labeled to have regression tests skipped?

**Review by pront — COMMENTED** [review]
/ci-run-regression

File: .github/workflows/regression.yml [hunk 6]
```diff
@@ -45,11 +47,30 @@ env:
...
+    runs-on: ubuntu-latest
+    outputs:
+      should-run: ${{ steps.set-should-run.outputs.SHOULD_RUN }}
+    steps:
+      - name: Determine if should run
+        id: set-should-run
+        run: |
+          if [[ "${{ github.event_name }}" == "pull_request" ]]; then
+            echo "SHOULD_RUN=false" >> $GITHUB_OUTPUT
+          elif [[ "${{ github.event_name }}" != "pull_request_review" &&
```
Comment (jszwedko) [inline .github/workflows/regression.yml:61]: Is this intended to match the `workflow_call`? Or are there any other events that should match?

**Review by pront — COMMENTED** [review]
/ci-run-regression

Edit:
Comment trigger worked -> https://github.com/vectordotdev/vector/actions/runs/11372868367

**Review by pront — COMMENTED** [review]
/ci-run-regression

Edit:
Added the label and the suite was triggered -> https://github.com/vectordotdev/vector/actions/runs/11373073641/job/31638883650

File: .github/workflows/regression.yml (hunk 6 above)
Comment (pront) [reply .github/workflows/regression.yml:61]: Thank you for this comment, it made me think a bit more about the whole check 🤔 

Updates in [this commit](https://github.com/vectordotdev/vector/pull/21507/commits/1547502b6987db491da34b8ae55d76cff99d6f03):
* Lifted the check `github.event_name != 'pull_request'` from the previously named job `should-run` (now called `check-source-changed`)
* Added check for the new label
  * caveat we still want this suite to be triggered by review comments hence the check for `pull_request_review`
* All other events fall under the `else` block, and the `should-run` output becomes `true`.

A couple of checks:
* https://github.com/vectordotdev/vector/pull/21507#pullrequestreview-2373537622
* https://github.com/vectordotdev/vector/pull/21507#pullrequestreview-2373568988
(I manually cancelled them after they progressed)

Comment (jszwedko) [reply .github/workflows/regression.yml:61]: I'm having a hard time following the conditional logic in here 😓 I think this might be more readable if we did something like (might contain typos):

```
if [[ "${{ github.event_name }}" == "pull_request" ]]; then
  echo "SHOULD_RUN=false" >> $GITHUB_OUTPUT
elif [[ "${{ github.event_name }}" == "pull_request_review" ]]; then
  echo "SHOULD_RUN=true" >> $GITHUB_OUTPUT
elif [[ "${{ contains(github.event.pull_request.labels.*.name, 'ci-condition: skip-regression') }}" == "true" ]]; then
  # The small caveat here is that we still want to run the suite via a review comment.
  echo "SHOULD_RUN=false" >> $GITHUB_OUTPUT
else
  echo "SHOULD_RUN=true" >> $GITHUB_OUTPUT
fi
```

I think these simplified expressions make it easier to see the intent.

**Comment by jszwedko** [comment]
Sorry, I accidentally unmerged this. I left one non-blocking nit here: https://github.com/vectordotdev/vector/pull/21507#discussion_r1803756641

File: .github/workflows/regression.yml (hunk 6 above)
Comment (pront) [reply .github/workflows/regression.yml:61]: I saw this just now after merging. I think these are equivalent. I wanted to see how this behaves for now dependabot PRs. I can follow-up to clean it up even more.

One question, for the following:
```
elif [[ "${{ github.event_name }}" == "pull_request_review" ]]; then echo "SHOULD_RUN=true" >> $GITHUB_OUTPUT
```

I suppose we will never hit this unless this was triggered by a `ci-run-regression` comment?

**Comment by pront** [comment]
Nice, the logic worked on this [dependabot PR](https://github.com/vectordotdev/vector/pull/21537): https://github.com/vectordotdev/vector/actions/runs/11388243430/job/31684731637

File: .github/workflows/regression.yml (hunk 6 above)
Comment (jszwedko) [reply .github/workflows/regression.yml:61]: Agreed, I think they are equivalent, I just find it hard to understand the current conditions 😓 



> One question, for the following:
> 
> ```
> elif [[ "${{ github.event_name }}" == "pull_request_review" ]]; then echo "SHOULD_RUN=true" >> $GITHUB_OUTPUT
> ```
> 
> I suppose we will never hit this unless this was triggered by a `ci-run-regression` comment?

Actually, I think `pull_request_review` will never be the trigger. Instead, it'll be `workflow_call` (where the caller, [ci-review.yml](https://github.com/vectordotdev/vector/blob/master/.github/workflows/ci-review-trigger.yml) has the `pull_request_review` trigger). So instead this should be:

```
elsif [[ "${{ github.event_name }}" == "workflow_call" ]]; then echo "SHOULD_RUN=true" >> $GITHUB_OUTPUT
```

Comment (pront) [reply .github/workflows/regression.yml:61]: Closing the loop: https://github.com/vectordotdev/vector/pull/21540
